include/aos/common/crypto/openssl/cryptoprovider.hpp

```diff
@@ -0,0 +1,229 @@
+     * @param[out] result DN text representation.
+     * @result Error.
+     */
+    Error ASN1DecodeDN(const Array<uint8_t>& dn, String& result) override;
+
+    /**
+     * Encodes array of object identifiers into ASN1 value.
+     *
+     * @param src array of object identifiers.
+     * @param asn1Value result ASN1 value.
+     */
+    Error ASN1EncodeObjectIds(const Array<asn1::ObjectIdentifier>& src, Array<uint8_t>& asn1Value) override;
+
+    /**
+     * Encodes big integer in ASN1 format.
+     *
+     * @param number big integer.
+     * @param[out] asn1Value result ASN1 value.
+     * @result Error.
+     */
+    Error ASN1EncodeBigInt(const Array<uint8_t>& number, Array<uint8_t>& asn1Value) override;
+
+    /**
+     * Creates ASN1 sequence from already encoded DER items.
+     *
+     * @param items DER encoded items.
+     * @param[out] asn1Value result ASN1 value.
+     * @result Error.
+     */
+    Error ASN1EncodeDERSequence(const Array<Array<uint8_t>>& items, Array<uint8_t>& asn1Value) override;
+
+    /**
+     * Returns value of the input ASN1 OCTETSTRING.
+     *
+     * @param src DER encoded OCTETSTRING value.
+     * @param[out] dst decoded value.
+     * @result Error.
+     */
+    Error ASN1DecodeOctetString(const Array<uint8_t>& src, Array<uint8_t>& dst) override;
+
+    /**
+     * Decodes input ASN1 OID value.
+     *
+     * @param inOID input ASN1 OID value.
+     * @param[out] dst decoded value.
+     * @result Error.
+     */
+    Error ASN1DecodeOID(const Array<uint8_t>& inOID, Array<uint8_t>& dst) override;
+
+    /**
+     * Creates UUID version 5 based on a given namespace identifier and name.
+     *
+     * @param space namespace identifier.
+     * @param name name.
+     * @result RetWithError<uuid::UUID>.
+     */
+    RetWithError<uuid::UUID> CreateUUIDv5(const uuid::UUID& space, const Array<uint8_t>& name) override;
+
+    /**
+     * Creates hash instance.
+     *
+     * @param algorithm hash algorithm.
+     * @return RetWithError<UniquePtr<HashItf>>.
+     */
+    RetWithError<UniquePtr<HashItf>> CreateHash(Hash algorithm) override;
+
+    /**
+     * Generates random integer value in range [0..maxValue].
+     *
+     * @param maxValue maximum value.
+     * @return RetWithError<uint64_t>.
+     */
+    RetWithError<uint64_t> RandInt(uint64_t maxValue) override;
+
+    /**
+     * Generates random buffer.
+     *
+     * @param[out] buffer result buffer.
+     * @param size buffer size.
+     * @return Error.
+     */
+    Error RandBuffer(Array<uint8_t>& buffer, size_t size = 0) override;
+
+private:
+    class OpenSSLHash : public crypto::HashItf, private NonCopyable {
+    public:
+        Error Init(ossl_lib_ctx_st* libCtx, const char* mdtype);
+        Error Update(const Array<uint8_t>& data) override;
+        Error Finalize(Array<uint8_t>& hash) override;
+        ~OpenSSLHash();
+
+    private:
+        evp_md_ctx_st* mMDCtx = nullptr;
```

Comment:
[optional] Will it work this way, so no need to add these structs forward declarations
```suggestion
        struct evp_md_ctx_st* mMDCtx = nullptr;
```

Reply:
thank you, fixed.